src/apps/vpn/connectionmanager.h

```diff
@@ -0,0 +1,244 @@
+                     uint64_t rxBytes);
+  void implInitialized(bool status, bool connected,
+                       const QDateTime& connectionDate);
+
+ signals:
+  void stateChanged();
+  void timeChanged();
+  void enableDisconnectInConfirmingChanged();
+  void handshakeFailed(const QString& serverHostname);
+  void connectionRetryChanged();
+  void newConnectionSucceeded();
+  void controllerDisconnected();
+  void readyToQuit();
+  void readyToUpdate();
+  void readyToBackendFailure();
+  void readyToServerUnavailable(bool pingReceived);
+  void activationBlockedForCaptivePortal();
+
+ public:
+  ConnectionManager();
+  ~ConnectionManager();
+
+  void initialize();
+  static QList<IPAddress> getAllowedIPAddressRanges(const Server& server);
+
+  enum ServerCoolDownPolicyForSilentSwitch {
+    eServerCoolDownNeeded,
+    eServerCoolDownNotNeeded,
+  };
+
+  bool silentSwitchServers(
+      ServerCoolDownPolicyForSilentSwitch serverCoolDownPolicy);
+
+ private:
+  enum NextStep {
+    None,
+    Quit,
+    Update,
+    Disconnect,
+    BackendFailure,
+    ServerUnavailable,
+  };
+
+  NextStep m_nextStep = None;
+
+  enum DNSPortPolicy {
+    ForceDNSPort,
+    DoNotForceDNSPort,
+  };
+
+  void activateInternal(DNSPortPolicy dnsPort,
+                        ServerSelectionPolicy serverSelectionPolicy);
+
+  void clearConnectedTime();
+  void clearRetryCounter();
+  QStringList getExcludedAddresses();
+  void activateNext();
+  void setState(State state);
+  void resetConnectedTime();
+  bool processNextStep();
+  void maybeEnableDisconnectInConfirming();
+  void serverDataChanged();
+
+ private:
+  QTimer m_timer;
+  QTimer m_connectingTimer;
+  QTimer m_handshakeTimer;
+
+  QDateTime m_connectedTimeInUTC;
+
+  State m_state = StateInitializing;
+  bool m_enableDisconnectInConfirming = false;
+  QList<InterfaceConfig> m_activationQueue;
+  int m_connectionRetry = 0;
+
+  QScopedPointer<ControllerImpl> m_impl;
+  bool m_portalDetected = false;
+
+  // Server data can change while the controller is busy completing an
+  // activation or a server switch because they are managed by the
+  // SettingsHolder object and exposed to user interaction, addons, and JS.
+  //
+  // But the controller needs to know the location to use for the entire
+  // duration of its tasks. When the client schedules a VPN activation,
+  // `m_serverData` is set as a copy of the current `MozillaVPN::serverData()`,
+  // ignoring further updates until the pending operations terminate. Instead,
+  // `m_nextServerData` is set when a server-switch request is scheduled while
+  // an activation operation is still in progress.
+  //
+  // At initialization time, these two member variables are set to
+  // MozillaVPN::serverData() to do not let not initialize.
+  //
+  // Please, do not use MozillaVPN::serverData() in the controller!
+  ServerData m_serverData;
+  ServerData m_nextServerData;
+
+  PingHelper m_ping_canary;
+  bool m_ping_received = false;
```

Comment:
Consider using lower camel casing for the names, for consistency with the C++ code base, now that it has moved to a new file.